core/src/main/java/io/aiven/kafka/tieredstorage/chunkmanager/CachingChunkManager.java

```diff
@@ -0,0 +1,132 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata;
+
+import io.aiven.kafka.tieredstorage.manifest.SegmentManifest;
+import io.aiven.kafka.tieredstorage.storage.StorageBackendException;
+
+import com.github.benmanes.caffeine.cache.AsyncCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.github.benmanes.caffeine.cache.Scheduler;
+import com.github.benmanes.caffeine.cache.Weigher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class CachingChunkManager<T> implements Configurable, ChunkManager {
+    private static final Logger log = LoggerFactory.getLogger(CachingChunkManager.class);
+    private static final long GET_TIMEOUT_SEC = 10;
+    private final ChunkManager chunkManager;
+    private final Executor executor;
+    protected AsyncCache<ChunkKey, T> cache;
+
+    public CachingChunkManager(final ChunkManager chunkManager, final Executor executor) {
+        this.chunkManager = chunkManager;
+        this.executor = executor;
+    }
+
+    /**
+     * Fetches a specific chunk from remote storage and stores into the cache.
+     * Since it's not possible to cache an opened InputStream the actual data is cached and everytime
+     * there is a call to cache the InputStream is recreated from the data stored in cache and stored into local
+     * variable. This also allows to solve the race condition between eviction and fetching. Since the InputStream is
+     * opened right when fetching from cache happens even if the actual value will be removed from the cache
+     * the InputStream will still contain the data.
+     */
+    public InputStream getChunk(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                final SegmentManifest manifest,
+                                final int chunkId) throws StorageBackendException, IOException {
+        final Uuid id = remoteLogSegmentMetadata.remoteLogSegmentId().id();
+        final ChunkKey chunkKey = new ChunkKey(id, chunkId);
+        final AtomicReference<InputStream> result = new AtomicReference<>();
+        try {
+            return cache.asMap().compute(chunkKey, (key, val) -> CompletableFuture.supplyAsync(() -> {
+                if (val == null) {
+                    try {
+                        final InputStream chunk = chunkManager.getChunk(remoteLogSegmentMetadata, manifest, chunkId);
+                        final T t = this.cacheChunk(chunkKey, chunk);
```

Comment:
Maybe I'm missing something also but is not it so that the actual reading from S3 will happen after the addition of the CompletableFuture into the cache? Inside `compute` method in `ConcurrentHashMap` the remapping function will be called  but it just creates the instance of CompletableFuture that will be run async because of `supplyAsync` call.

Reply:
Yep, you're right. Sorry, my bad.